File: app/src/main/java/com/arv/app/core/data/Invitation.kt
package com.arv.app.core.data

import com.arv.app.core.data.local.InviteEntity
import com.arv.app.core.data.local.MemberEntity

/**
 * What happens when somebody types a code in.
 *
 * Pure so it can be tested without a database, same as [Membership]. Every refusal is a
 * distinct answer rather than one "invalid code", because the person standing there needs
 * to know whether to retype it, ask for a new one, or stop trying. "Invalid" tells them
 * none of that and sends them back to the person who invited them with nothing useful.
 */
object Invitation {

    sealed interface Result {
        /** Good code. [member] is the row to write, [spent] is the invite to close out. */
        data class Accepted(val member: MemberEntity, val spent: InviteEntity) : Result

        /** Not shaped like a code at all. They are still typing, or they mistyped badly. */
        data object NotACode : Result

        /** Shaped right, but no such invitation. A typo that happens to be well formed. */
        data object Unknown : Result

        /**
         * Already spent. Distinct from [Unknown] on purpose: this one means "ask them for
         * a fresh code", which is a thing the person can actually act on.
         */
        data object AlreadyUsed : Result

        /** The person who issued it took it back before anyone used it. */
        data object Revoked : Result

        /** Their own code. Nobody invites themselves into a family. */
        data object YourOwn : Result

        /** They are already in. Not an error, just nothing to do. */
        data object AlreadyInThisFamily : Result
    }

    fun redeem(
        typed: String?,
        /** The row matching [typed] once normalized, or null if there wasn't one. */
        invite: InviteEntity?,
        /** This account's existing standing in [InviteEntity.familyId], if any. */
        existingMember: MemberEntity?,
        userId: String,
        nowMillis: Long
    ): Result {
        InviteCode.normalize(typed) ?: return Result.NotACode
        if (invite == null) return Result.Unknown
        if (invite.revokedAt != null) return Result.Revoked
        if (invite.usedAt != null) return Result.AlreadyUsed
        if (invite.issuedByUserId == userId) return Result.YourOwn
        if (existingMember != null) return Result.AlreadyInThisFamily

        return Result.Accepted(
            member = MemberEntity(
                familyId = invite.familyId,
                userId = userId,
                role = invite.grantsRole,
                // The whole reason the code is per person. Written once, at the moment it
                // is true, and never inferred afterwards.
                invitedBy = invite.issuedByUserId,
                joinedAt = nowMillis
            ),
            spent = invite.copy(usedAt = nowMillis, usedByUserId = userId)
        )
    }
}

File: functions/invitation.js
// The decision, kept pure so it can be tested without Firestore and read next to
// app/src/main/java/com/arv/app/core/data/Invitation.kt, which it mirrors line for line
// and in the same order. If the two ever disagree, a code is accepted on one phone and
// refused on another, and the person holding it has no way to know which to believe.

/** Deliberately missing 0, O, 1, I and L. See InviteCode.kt for why. */
const ALPHABET = "ABCDEFGHJKMNPQRSTUVWXYZ23456789";
const LENGTH = 6;
const CANONICAL = new RegExp(`^[${ALPHABET}]{${LENGTH}}$`);

/**
 * What the person typed, turned into what we can look up, or null if it could never be a
 * code. Case and separators are ignored on the way in, because the person typing it did
 * not choose how it was written down.
 */
function normalize(input) {
  if (typeof input !== "string") return null;
  const stripped = input.toUpperCase().replace(/[^A-Z0-9]/g, "");
  return CANONICAL.test(stripped) ? stripped : null;
}

/**
 * Seven answers, never "invalid". Each one tells the person what to do next.
 *
 * Withdrawn and spent are checked before expiry on purpose: those name something a person
 * did, and saying which is more use than saying time passed.
 *
 * @param {object|null} invite    the invite document, or null when no such code exists
 * @param {boolean} alreadyMember whether this account already has a standing in the family
 * @param {string} uid            the account redeeming
 * @param {number} nowMillis      the clock, passed in so the decision is testable
 */
function decide(invite, alreadyMember, uid, nowMillis) {
  if (!invite) return { status: "UNKNOWN" };
  if (invite.revokedAt != null) return { status: "REVOKED" };
  if (invite.usedAt != null) return { status: "ALREADY_USED" };
  if (invite.expiresAt != null && nowMillis >= invite.expiresAt) return { status: "EXPIRED" };
  if (invite.createdBy === uid) return { status: "YOUR_OWN" };
  if (alreadyMember) return { status: "ALREADY_IN_THIS_FAMILY" };
  return { status: "ACCEPTED" };
}

module.exports = { normalize, decide, ALPHABET, LENGTH };
